expat/tests/common.c

```diff
@@ -181,18 +181,18 @@ enum XML_Status
 _XML_Parse_SINGLE_BYTES(XML_Parser parser, const char *s, int len,
                         int isFinal) {
   const int chunksize = g_chunkSize;
-  int offset = 0;
   if (chunksize > 0) {
     // parse in chunks of `chunksize` bytes as long as possible
-    for (; offset + chunksize < len; offset += chunksize) {
-      enum XML_Status res = XML_Parse(parser, s + offset, chunksize, XML_FALSE);
+    for (; chunksize <= len;
+         len -= chunksize, s = (s != NULL) ? (s + chunksize) : NULL) {
```

Comment:
The NULL condition is unnecessary: chunksize is known to be non-zero here (we're only concerned about `NULL+0`, right?).

(also, len is known to be non-zero, and if someone sends in `s=NULL` and `len>0`, they _deserve_ the UB error.)

Reply:
> The NULL condition is unnecessary: chunksize is known to be non-zero here (we're only concerned about `NULL+0`, right?).

@Snild-Sony my patch was done with the deliberate "requirement" (or assumption) in mind to make `_XML_Parse_SINGLE_BYTES` support the full interface of `XML_Parse` but with chunking support added — a drop-in replacement — so that eventually we can replace _all_ calls to `XML_Parse` by `_XML_Parse_SINGLE_BYTES` except those two inside `_XML_Parse_SINGLE_BYTES`. The tests in `test_empty_parse` below are the ones that challenge that requirement and triggered undefined behavior:

https://github.com/libexpat/libexpat/blob/3a0c5d6ae53d64fd8f8abc0ff03dcd6cfb754030/expat/tests/basic_tests.c#L2730-L2758

And these include `NULL`. Of course we could exclude those cases but I would vote for making `_XML_Parse_SINGLE_BYTES` robust enough to handle all and then we don't need to check again for leftovers or document why particular calls are exempt. But maybe that's just my preference.

If you only concern left is to resolve `(s != NULL) ? x : y` to `x` then I would want to vote to leave that part in and go forward. One alternative would be to revert the whole patch to `_XML_Parse_SINGLE_BYTES` and make test `test_empty_parse` exercise plain `XML_Parse`. Both can be argued for.

What do you think?

Reply:
> > The NULL condition is unnecessary: chunksize is known to be non-zero here (we're only concerned about `NULL+0`, right?).
> 
> @Snild-Sony my patch was done with the deliberate "requirement" (or assumption) in mind to make `_XML_Parse_SINGLE_BYTES` support the full interface of `XML_Parse` but with chunking support added — a drop-in replacement — so that eventually we can replace _all_ calls to `XML_Parse` by `_XML_Parse_SINGLE_BYTES`

I was under the impression that we were only trying to avoid adding `0`, since that was the specific error we had. My bad.

However, my other point about len=0 making sure we don't reach that code still stands:

> The tests in test_empty_parse below are the ones that challenge that requirement and triggered undefined behavior:  [...] And these include NULL.

But all of those also have len=0, which means the execution would not reach the line of code that we're discussing.

Reply:
> If you only concern left is to resolve `(s != NULL) ? x : y` to `x`

It is. 

> then I would want to vote to leave that part in and go forward.

Making that choice is your privilege and your burden as maintainer. My words hold no power here. :)

Reply:
> > If you only concern left is to resolve `(s != NULL) ? x : y` to `x`
> 
> It is.

@Snild-Sony I think I may have found (and pushed) a compromise that gets rid of that and still doesn't do undefined behavior:
https://github.com/libexpat/libexpat/compare/421551302a37ddbdf7874c877070df2fa4d79567..6b5691bcf8b93668d8733aa3e99510be6e717a3b
Any better? What do you think?

> > then I would want to vote to leave that part in and go forward.
> 
> Making that choice is your privilege and your burden as maintainer. My words hold no power here. :)

Let's try with consensus first. Maybe we're already close to consensus now.

Reply:
That part looks good now, thanks.